internal/storage/inmem/store.go

```diff
@@ -0,0 +1,328 @@
+package inmem
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/go-memdb"
+
+	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/internal/storage"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// Store implements an in-memory resource database using go-memdb.
+//
+// It can be used as a storage backend directly via the Backend type in this
+// package, but also handles reads in our Raft backend, and can be used as a
+// local cache when storing data in external systems (e.g. RDBMS, K/V stores).
+type Store struct {
+	db  *memdb.MemDB
+	pub *stream.EventPublisher
+}
+
+// NewStore creates a Store.
+//
+// You must call Run before using the store.
+func NewStore() (*Store, error) {
+	db, err := memdb.NewMemDB(&memdb.DBSchema{
+		Tables: map[string]*memdb.TableSchema{
+			tableNameMetadata: {
+				Name: tableNameMetadata,
+				Indexes: map[string]*memdb.IndexSchema{
+					indexNameID: {
+						Name:         indexNameID,
+						AllowMissing: false,
+						Unique:       true,
+						Indexer:      &memdb.StringFieldIndex{Field: "Key"},
+					},
+				},
+			},
+			tableNameResources: {
+				Name: tableNameResources,
+				Indexes: map[string]*memdb.IndexSchema{
+					indexNameID: {
+						Name:         indexNameID,
+						AllowMissing: false,
+						Unique:       true,
+						Indexer:      idIndexer{},
+					},
+					indexNameOwner: {
+						Name:         indexNameOwner,
+						AllowMissing: true,
+						Unique:       false,
+						Indexer:      ownerIndexer{},
+					},
+				},
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	s := &Store{
+		db:  db,
+		pub: stream.NewEventPublisher(10 * time.Second),
+	}
+	s.pub.RegisterHandler(eventTopic, s.watchSnapshot, false)
+
+	return s, nil
+}
+
+// Run until the given context is canceled. This method blocks, so should be
+// called in a goroutine.
+func (s *Store) Run(ctx context.Context) { s.pub.Run(ctx) }
+
+// Read a resource using its ID.
+//
+// For more information, see the storage.Backend documentation.
+func (s *Store) Read(id *pbresource.ID) (*pbresource.Resource, error) {
+	tx := s.db.Txn(false)
+	defer tx.Abort()
+
+	val, err := tx.First(tableNameResources, indexNameID, id)
+	if err != nil {
+		return nil, err
+	}
+	if val == nil {
+		return nil, storage.ErrNotFound
+	}
+
+	res := val.(*pbresource.Resource)
+
+	// Observe the Uid if it was given.
+	if id.Uid != "" && res.Id.Uid != id.Uid {
+		return nil, storage.ErrNotFound
+	}
+
+	// Let the caller know they need to upgrade/downgrade the schema version.
+	if id.Type.GroupVersion != res.Id.Type.GroupVersion {
+		return nil, storage.GroupVersionMismatchError{
+			RequestedType: id.Type,
+			Stored:        res,
+		}
+	}
+
+	return res, nil
+}
+
+// WriteCAS performs an atomic Check-And-Set (CAS) write of a resource.
+//
+// For more information, see the storage.Backend documentation.
+func (s *Store) WriteCAS(res *pbresource.Resource, vsn string) error {
```

Comment:
Is it expected that the version of the resource is set on the resource prior to invoking `WriteCAS` and `DeleteCAS`?

Should these two methods additionally take in the new version and update the resource accordingly?

Reply:
Yep, exactly. The expectation is that the Resource Service will generate the `Version`, `Generation`, and `Uid` so we can use a consistent scheme ([ULIDs](https://github.com/ulid/spec)) across the board.

The storage backend doesn't need to do anything special to update the version (or even validate that it has been changed) so I don't think we need a separate argument.

Reply:
I am not sure ULIDs are appropriate for the resource version (generation and uid should be fine).

For a single resource I think ULIDs can provide write linearizability as we generally expect in a few places.

Scenario (two concurrent writes to a single resource):
* Some resource exists with ULID version of `01GVJVKJTMXSD3FT9ZSEGJEAVM`
* Write A received
   * ULID generated is `01GVJVMFHBT9NK7MYEXCMCNFHG`
   * CAS version to be used will be `01GVJVKJTMXSD3FT9ZSEGJEAVM`
* Write B received 
   * ULID generated is `01GVJVMFHBZC4SWHT5XPCZJNWW`
   * CAS version to be used will be `01GVJVKJTMXSD3FT9ZSEGJEAVM`
* _Note that writes A an B share the same 10 byte timestamp prefix due to being generated in the same millisecond. Also note that write Bs ULID is lexicographically after A_
* _For whatever reason the validation callbacks for write A take longer than write B_
* API sends write B to the storage backend
* Storage backend applies write B as the CAS index matches the resources current version.
* API sends write A to the storage backend
* Storage backend rejects write A due to the CAS index mismatch since write B changed the resource version
* API retries write A
   * New ULID is generated (_I think this must happen in order for the ULID to be lexicographically after write B, however that will only be true if the process to attempt writing to the storage backend and be rejected would take more than 1ms_)
   * CAS version is updated to write Bs version
* Storage Backend applies write A

I think this is okay because:

1. We are CASifying all writes.
2. It is practically impossible today for the time between ULID generation and ULID regeneration after a rejected CAS write to take less than 1ms.

The second constraint there I think will be valid for the Raft backend or any generally any backend that does disk or network IO (like k8s or dynamo or even some flat file backend). However for the inmem storage backend for testing I am not quite as convinced that it will always be the case.

Things get even gnarlier when you consider operations against multiple resources of some kind such as a resource listing. Since concurrent writes to multiple distinct resources will not have CAS operations to help enforce version linearizability it is entirely possible and even expected for resources to be written and events seen for resources out of lexicographic order based on the ULID version.

Is this a problem for watches, particularly if we want to allow watch resumption? What about if we wanted to support [functionality similar to k8s](https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions) where you can provide a resource version to list operations and only get data newer than that version. 

In some of those cases strict linearizability of all resource writes of a particular type is necessary. Do we care and are we willing to rule out implementing those features to avoid having the storage backends more involved in setting resource versions?

Reply:
I probably should've clarified, I'm not planning to depend on the lexicographical sortability of ULIDs for determining the order of events or anything else related to correctness. The timestamp component is inherently susceptible to clock skew across servers, and other anomalies on the same server, so this wouldn't be safe, even if the "entropy" component was a monotonic index.

Essentially, we'll be treating ULIDs as opaque random strings, but with the benefit of being able to tell if one version came **roughly** before or after another (think day/week/month level of granularity) 😄 but if this leads to folks making greater assumptions about ordering, I'd be keen to replace it with a truly random string.

I haven't accounted for watch resumption, or listing resources that are "newer" than a given version, because they require such strict consistency guarantees ([serializability](https://jepsen.io/consistency/models/serializable) I think) to implement correctly, which would rule out a lot of potential external storage systems.

I guess what you're saying is that Raft *does* provide serializability, so why not use the log index as a version, so that we at least have the option of supporting these features in the future? I'm not strongly against this, but would argue that we should think carefully before adding features that can only be satisfied with strong consistency, because it closes the door on lots of potentially great storage integrations.

Reply:
Following up from our call earlier:

We're going to make storage backends responsible for managing the version, and have the Raft backend use the log index so that we can build a blocking query layer on top (again, for backward-compatibility with the v1 API). Other backends will probably continue to treat versions as opaque strings, and make them ULIDs or something.